### ISSUE TITLE
Possibility of filtering `property<[foo]>` before using for loop

### ISSUE THREAD

**Issue by dilawar** [slint-ui/slint#1328]
I have a `property<[FooMsg]> msgs` that stores list messages I want to display in different tabs. I am not maintaining different properties for each tab. I can do that if the answer here is a No. Each message `FooMsg` has a field called `tag` that determines the `Tab` of the message. 

Inside a `Tab`,  Here is a psedo code that I use to filter the messages and display them.

```
for msg[i] in root.msgs {
   if msg.tag == root.tag { 
      Text {
         text : msg.text;
         row: i;
       }
   }
}
```
I was hoping for something like (pseudo code):

```
for msg in root.msg.filter( |x| x.tag == root.tag) {
   Text { text: msg.text, row : i };
}
```
Note that the value of `i` will be different in the second case.  The second case is suitable for generating a `GridLayout` because `i` will take values that increase by 1 at each step. 

In short, I am looking for a `filter` like function for the array before I can iterate over it. Suggestions?

**Comment by tronical** [comment]
While this isn’t possible in pure Slint yet, you can do this if you provide your Mode from Rust and place it behind a [FilterModel](https://slint-ui.com/snapshots/master/docs/rust/slint/struct.filtermodel). Would that help?

**Comment by tronical** [comment]
FWIW it would be nice to be able to write a filter like in your example in pure Slint as well.

**Comment by ogoffart** [comment]
We just would need to find a good syntax for it.

```rust
// Like in rust
for xx in model.filter(|x| x.tag == root.tag) : Rectangle {
// like in JS
for xx in model.filter(x => x.tag == root.tag) : Rectangle {
// specialized syntax
for xx in model | #.tag == root.tag : Rectangle {
```

We also would probably need a `map` adapror syntax as well.  
Maybe a sorting one too?

**Comment by tronical** [comment]
Right, and we'd need the map/filter/sort in C++ as well. This sounds very well doable. My order of preference in syntax is JS > Rust > Special ;-)

**Comment by dilawar** [comment]
> While this isn’t possible in pure Slint yet, you can do this if you provide your Mode from Rust and place it behind a [FilterModel](https://slint-ui.com/snapshots/master/docs/rust/slint/struct.filtermodel). Would that help?

Thanks! This might just work 🥳 .

**Comment by dilawar** [comment]
I also think the JS syntax is easier on the eyes than Rust syntax (especially when you are also supporting JS and C++).

**Comment by dilawar** [comment]
> While this isn’t possible in pure Slint yet, you can do this if you provide your Mode from Rust and place it behind a [FilterModel](https://slint-ui.com/snapshots/master/docs/rust/slint/struct.filtermodel). Would that help?

Thanks again for the suggestion @tronical. It worked! Not sure if I am using it the most efficiently but for now it solves my problem. 

```Rust
        // Connect msglist to slint Window.
        let msglist_app = ModelRc::new(FilterModel::new(self.msglist.clone(), |s| s.code == "App"));
        self.window.unwrap().set_msglist(msglist_app);
```

**Comment by tronical** [comment]
👍 We could leave this ticket open to track the ability to specify filters in .slint itself. If you're okay with that then I'll edit the title slightly to reflect the intent.

**Comment by dilawar** [comment]
@tronical Sounds good.

**Comment by jmgrosen** [comment]
Any other workarounds suggested for this? It seems like it makes maintaining the "one source of truth" that Slint seems to otherwise do pretty well break quickly without losing modularity.

**Comment by quiettimejsg** [comment]
Issue Response: Model Filtering and Mapping Implementation Progress
Subject: Progress Update: Slint Model Filtering and Mapping Implementation

Dear Team,

I've completed the initial implementation of two core model adapters as part of our model enhancement initiative:

FilteredModel: Reactive row filtering with incremental index updates

MappedModel: Data transformation with LRU caching

Due to time constraints and complexity considerations, the following items from the original roadmap remain pending:

Compiler integration for .filter()/.map() syntax

Advanced sorting adapter implementation

Performance benchmarking infrastructure

Full documentation generation

The completed components are production-ready and include:
✅ Incremental filtering index updates (O(log n) performance)
✅ Bidirectional data mapping with error handling
✅ LRU caching with size monitoring
✅ Comprehensive event propagation system
✅ Panic-safe mapping functions
✅ Thread safety enforcement (!Send/!Sync)

I've created PR #8782 with the initial implementation for review.

**Comment by ogoffart** [comment]
This issue is for the implementation of something like `.map(x => {...})` and `.filter(x => {...})`  in Slint code.

There are a few part to this:

1. Parsing of the lambda construct (in the parser.rs module)
2. Semantic analysis to make sure that everything is correct. (in the resolving.rs) In particular that the closure can only access what is going to be accessible. And have error when this is not the case.
3. Interpreter and Code generation in Rust/C++, this would be using the existing MapModel/FilterModel

The step 2 might be a bit tricky as we need to make sure everything is sound.

Consider this:

```slint
component Foo {
    in property <bool> condition;
    in property <[XXX]> model1;
    out property <[XXX]> model2;
    
    if condition : bar := Bar {
        model: model1.filter(x => {x.foo == bar.value});
        frobnicate => { root.model2 = bar.model; }
    }
}
```

When `bar.value` changes, we expect the `model2` to change as well.
But when `condition` becomes false, `bar` gets destroyed and the model2 would still reference bar.  The user of Foo might have copied model2 anywhere. So we need to define a semantic. SOme options:
  - We don't allow accessing properties from the lambda expression. But this might be a bit restricting
  - Or we make a copy of the value so this would be the same as `let value = bar.value; model1.filter(x => x.foo == value))`.  But that way, when `bar.value` changes, model2 doesn't change.
  - Or we keep a weak pointer to the item tree in the capture. Meaning that when `bar` gets destroyed, bar.value would return to a default-contructed type, or the whole expression would return a default constructed value.
  - Or we use a concept of live time to figure out what access are possible (really really hard)

**Comment by ubruhin** [comment]
Would it be an option to just detect if a filter expression accesses properties, and then disallow copying such a model? E.g.

```slint
component Foo {
    in property <bool> condition;
    in property <[XXX]> model1;
    out property <[XXX]> model2;
    
    if condition : bar := Bar {
        model-a: model1.filter(x => {x.foo == "literal"});
        model-b: model1.filter(x => {x.foo == bar.value});

        frobnicate => { 
            root.model2 = bar.model-a;  // Works
            root.model2 = bar.model-b;  // Compile error "attempting to copy model with dynamic filter"
        }
    }
}
```

Or alternatively, disable copying filtered models completely? I could imagine copying models is not done very often and thus not a big restriction 🤔 But I don't know how feasible it is to disallow copying models and if this would really solve all problems.

**Comment by tgrushka** [comment]
The lack of ability to map a display property in a widget as simple as a ComboBox without having to define additional Slint-specific models in my Rust code (and thereby mix UI code where it does not belong), to work around a limitation that does not exist in other frameworks, has forced me to abandon Slint in my project. 3.5 years of stagnation seems to suggest the lack of this feature may be a philosophical decision of Slint rather than a technical backlog item, therefore I am going with a different framework. Too bad, given that nearly every popular UI framework supports such simple mapping.

Example of what I was hoping to do:
```slint
// Region
ComboBox {
    width: 150px;
    model: [
        for region[idx] in region-list {
            region.name,
        }
    ];  
    current-value: root.current-region-name;
}
```

Fails to compile with the following errors:
```
Syntax error: expected ']'
Syntax error: expected ';'
Parse error
```

Given that the following compiles perfectly for a more complex component, it makes no sense that this syntax works in custom delegates which are only available in complex widgets but fails to work in simpler widgets:
```slint

// Left panel - Region list
VerticalBox {
width: 400px;
spacing: 8px;

Text {
    text: "Regions";
    font-size: 16px;
    font-weight: 700;
    color: white;
}

StandardListView {
    height: 300px;
    for region[idx] in region-list: Rectangle {
        background: idx == current-region-index ? #3d3d3d : #2d2d2d;
        border-color: #555;
        border-width: 1px;
        height: 40px;

        HorizontalBox {
            padding: 8px;
            spacing: 8px;
            Text {
                text: region.name;
                color: white;
                width: 50%;
            }
            VerticalBox {
                Text {
                    text: "Source: " + region.start-tc;
                    font-size: 10px;
                    color: #aaa;
                }
                Rectangle {
                    background: #222;
                    LineEdit {
                        text: region.target-start-tc;
                        placeholder-text: "Target TC";
                    }
                }
            }
        }
    }
}
```

I hope this feedback inspires the Slint maintainers to reconsider this design limitation in light of real world use cases.
